cmd/zpool/zpool_main.c

```diff
@@ -2836,12 +3393,263 @@ get_timestamp_arg(char c)
+	return (0);
+}
+
+/*
+ * Newer versions of the zfs module report back a bitmask of all the stats
+ * it supports.  Return the bitmask of stats that are supported on all pools.
+ */
+static uint64_t
+get_stat_flags(zpool_list_t *list)
+{
+	uint64_t mask = -1; /* Start off with all 0xFFs */
+
+	/*
+	 * __get_stat_flags() will lop off bits from "mask" until only the
+	 * flags that are supported on all pools remain.
+	 */
+	pool_list_iter(list, B_FALSE, __get_stat_flags, &mask);
+	return (mask);
+}
+
+/* for_each_vdev() function prototype */
+typedef int (*pool_vdev_iter_f)(zpool_handle_t *, nvlist_t *, void *);
+
+static int
+__for_each_vdev(zpool_handle_t *zhp, nvlist_t *nv, pool_vdev_iter_f func,
+		void *data)
+{
+	nvlist_t **child;
+	uint_t c, children;
+	int ret = 0;
+	int i;
+	char *type;
+
+	const char *list[] = {
+		ZPOOL_CONFIG_SPARES,
+		ZPOOL_CONFIG_L2CACHE,
+		ZPOOL_CONFIG_CHILDREN
+	};
+
+	for (i = 0; i < ARRAY_SIZE(list); i++) {
+		if (nvlist_lookup_nvlist_array(nv, list[i],
+					&child, &children) == 0) {
+			for (c = 0; c < children; c++) {
+				uint64_t ishole = B_FALSE;
+
+				(void) nvlist_lookup_uint64(child[c],
+					ZPOOL_CONFIG_IS_HOLE,
+					&ishole);
+
+				if (ishole)
+					continue;
+
+				ret |= __for_each_vdev(zhp, child[c], func,
+							data);
+			}
+		}
+	}
+
+	if (nvlist_lookup_string(nv, ZPOOL_CONFIG_TYPE, &type) != 0)
+		goto end;
+
+	/* Don't run our function on root vdevs */
+	if (strcmp(type, VDEV_TYPE_ROOT) != 0) {
+		ret += func(zhp, nv, data);
+	}
+
+end:
+	return (ret);
+}
+
+/*
+ * This is the equivalent of for_all_pools() for vdevs.  It iterates thorough
+ * all vdevs in the pool, ignoring root vdevs and holes, calling func() on
+ * each one.
+ *
+ * @zhp:	Zpool handle
+ * @func:	Function to call on each vdev
+ * @data:	Custom data to pass to the function
+ */
+static int
+for_each_vdev(zpool_handle_t *zhp, pool_vdev_iter_f func, void *data)
+{
+	nvlist_t *config, *nvroot;
+
+	if ((config = zpool_get_config(zhp, NULL)) != NULL) {
+		verify(nvlist_lookup_nvlist(config, ZPOOL_CONFIG_VDEV_TREE,
+					&nvroot) == 0);
+	}
+	return (__for_each_vdev(zhp, nvroot, func, data));
+}
+
+/*
+ * Return 1 if cb_data->cb_vdev_names[0] is this vdev's name, 0 otherwise.
+ */
+static int
+__is_vdev(zpool_handle_t *zhp, nvlist_t *nv, void *cb_data)
+{
+	iostat_cbdata_t *cb = cb_data;
+	char *name;
+
+	name = zpool_vdev_name(g_zfs, zhp, nv, cb->cb_name_flags);
+
+	if (strcmp(name, cb->cb_vdev_names[0]) == 0)
+		return (1); /* match */
+
+	return (0);
+}
+
+/*
+ * Returns 1 if cb_data->cb_vdev_names[0] is a vdev name, 0 otherwise.
+ */
+static int
+is_vdev(zpool_handle_t *zhp, void *cb_data)
+{
+	return (for_each_vdev(zhp, __is_vdev, cb_data));
+}
+
+/*
+ * Check if vdevs are in a pool
+ *
+ * Return 1 if all argv[] strings are vdev names in pool "pool_name". Otherwise
+ * return 0.  If pool_name is NULL, then search all pools.
+ */
+static int
+are_vdevs_in_pool(int argc, char **argv, char *pool_name,
```

Comment:
Unfortunately in this case it's specific to `zpool_main.c`, since `__is_vdev()` expects "cb_data" to be an iostat_cbdata_t.